### PR TITLE
indexed job: remove pods with invalid index

### PR DESCRIPTION
**Pull request by AliceZhang2016**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
For indexed jobs, when the job adopts pods from old job with larger index range, those pods with index larger or equal to current .spec.completions should be discarded.

#### Which issue(s) this PR fixes:
Ref kubernetes/enhancements#2214

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by AliceZhang2016** [comment]
cc @alculquicondor

File: pkg/controller/job/indexed_job_utils.go [hunk 1]
```diff
@@ -51,6 +51,11 @@ func calculateSucceededIndexes(pods []*v1.Pod) (string, int32) {
 		if ix == unknownCompletionIndex {
 			continue
 		}
+		// avoids invalid succeeded pods be counted in
+		// when do the pod adoption from old job with the larger completions value
```
Comment (alculquicondor) [inline pkg/controller/job/indexed_job_utils.go:55]: no need for this comment.

File: pkg/controller/job/job_controller.go [hunk 2]
```diff
@@ -759,6 +759,18 @@ func (jm *Controller) manageJob(job *batch.Job, activePods []*v1.Pod, succeeded
 		return active, err
 	}
 
+	if isIndexedJob(job) {
+		// job which adopts pods from an old job with larger .spec.Completions
+		// should discard pods with invalid index
+		podsToDelete := activePodsWithInvalidIndexForRemoval(job, activePods, int(*job.Spec.Completions))
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:765]: better modify `activePodsForRemoval` to handle this (it already removes pods with negative index, for example).

File: pkg/controller/job/job_controller_test.go [hunk 3]
```diff
@@ -560,6 +560,44 @@ func TestControllerSyncJob(t *testing.T) {
 			expectedCreatedIndexes: sets.NewInt(3, 4),
 			indexedJobEnabled:      true,
 		},
+		"indexed job with running pod adoption from job with larger index range": {
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller_test.go:563]: you can just name it: `indexed job with indexes outside of range`.
Where those pods come from (like adoption) is irrelevant for the test.
I would combine running, succeeded and failed in a single test.

File: pkg/controller/job/job_controller_test.go [hunk 4]
```diff
@@ -560,6 +560,44 @@ func TestControllerSyncJob(t *testing.T) {
...
+			parallelism:    2,
+			completions:    5,
+			backoffLimit:   6,
+			completionMode: batch.IndexedCompletion,
+			podsWithIndexes: []indexPhase{
+				{"0", v1.PodSucceeded},
+				{"6", v1.PodRunning},
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller_test.go:570]: set it to the first that should not count: 5

File: pkg/controller/job/job_controller_test.go [hunk 5]
```diff
@@ -560,6 +560,44 @@ func TestControllerSyncJob(t *testing.T) {
...
+		},
+		"indexed job with succeeded pod adoption from job with larger index range": {
+			parallelism:    2,
+			completions:    5,
+			backoffLimit:   6,
+			completionMode: batch.IndexedCompletion,
+			podsWithIndexes: []indexPhase{
+				{"0", v1.PodSucceeded},
+				{"6", v1.PodSucceeded},
+				{"7", v1.PodSucceeded},
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller_test.go:590]: what about Failed pods with indexes outside the range? They shouldn't count as failures.

File: pkg/controller/job/job_controller.go [hunk 6]
```diff
@@ -900,6 +900,7 @@ func activePodsForRemoval(job *batch.Job, pods []*v1.Pod, rmAtLeast int) []*v1.P
 		rm = make([]*v1.Pod, 0, rmAtLeast)
 		left = make([]*v1.Pod, 0, len(pods)-rmAtLeast)
 		rm, left = appendDuplicatedIndexPodsForRemoval(rm, left, pods)
+		rm = appendPodsWithInvalidIndexForRemoval(rm, left, int(*job.Spec.Completions))
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:903]: we need to know which pods were not removed. Perhaps you can incorporate the logic in `appendDuplicatedIndexPodsForRemoval`

File: pkg/controller/job/job_controller.go [hunk 7]
```diff
@@ -953,7 +954,8 @@ func getBackoff(queue workqueue.RateLimitingInterface, key interface{}) time.Dur
 func countPodsByPhase(job *batch.Job, pods []*v1.Pod, phase v1.PodPhase) int {
 	result := 0
 	for _, p := range pods {
-		if phase == p.Status.Phase && (!isIndexedJob(job) || getCompletionIndex(p.Annotations) != unknownCompletionIndex) {
+		idx := getCompletionIndex(p.Annotations)
+		if phase == p.Status.Phase && (!isIndexedJob(job) || (idx != unknownCompletionIndex) && idx < int(*job.Spec.Completions)) {
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:958]: can you make the parenthesis like `a && (b || (c && d))`

File: pkg/controller/job/job_controller_test.go [hunk 8]
```diff
@@ -2168,8 +2193,15 @@ func checkJobCompletionEnvVariable(t *testing.T, spec *v1.PodSpec) {
 	}
 }
 
-func hasFailingPods(status []indexPhase) bool {
+func hasFailingPods(status []indexPhase, completions int) bool {
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller_test.go:2196]: maybe rename to `hasValidFailingPods` and explain in a comment that the job controller only cares about failed pods with indexes in the valid range.

File: pkg/controller/job/job_controller_test.go [hunk 9]
```diff
@@ -691,7 +716,7 @@ func TestControllerSyncJob(t *testing.T) {
 				if err == nil {
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller_test.go:716]: I'm starting to dislike all this logic to determine if there should have been an error or not.

We should rather add a field to the test table `expectError bool` and perhaps put a comment on why we expect an error or not for each test case. Could be for a follow up PR though.

Comment (AliceZhang2016) [reply pkg/controller/job/job_controller_test.go:716]: Agree. maybe also add error message to the test table when there should be an error?

Comment (alculquicondor) [reply pkg/controller/job/job_controller_test.go:716]: Ideally, we should match the error that we expect, but perhaps a string comparisons is fine to start

Comment (AliceZhang2016) [reply pkg/controller/job/job_controller_test.go:716]: Then I will start another PR to modify the error match logic

File: pkg/controller/job/indexed_job_utils_test.go [hunk 10]
```diff
@@ -199,6 +223,7 @@ func TestAppendDuplicatedIndexPodsForRemoval(t *testing.T) {
 				{"3", v1.PodRunning},
 				{"5", v1.PodPending},
 			},
+			completions: 6,
 		},
 		"mixed": {
 			pods: []indexPhase{
```
Comment (alculquicondor) [inline pkg/controller/job/indexed_job_utils_test.go:229]: add some outside of the range

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/assign @soltysh

**Comment by nikhita** [comment]
Removing and re-adding the sig label, so that prow auto-removes the `do-not-merge/needs-sig` label

/remove-sig apps

**Comment by AliceZhang2016** [comment]
@janetkuo Could you please take a look at this PR to see if it can be approved? Thanks a lot!

**Review by janetkuo — APPROVED** [review]

**Pull request merged**